src/librustc_typeck/check/mod.rs

```diff
@@ -4787,34 +4818,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     }
                     ty::ImplContainer(_) => {}
                 }
-
-                let generics = self.tcx.generics_of(def_id);
-                if segments.len() >= 2 {
-                    let parent_generics = self.tcx.generics_of(generics.parent.unwrap());
-                    type_segment = Some((&segments[segments.len() - 2], parent_generics));
-                } else {
+                if segments.len() == 1 {
                     // `<T>::assoc` will end up here, and so can `T::assoc`.
                     let self_ty = opt_self_ty.expect("UFCS sugared assoc missing Self");
                     ufcs_associated = Some((container, self_ty));
                 }
-                fn_segment = Some((segments.last().unwrap(), generics));
             }
-
-            // Case 4. Local variable, no generics.
-            Def::Local(..) | Def::Upvar(..) => {}
-
-            _ => bug!("unexpected definition: {:?}", def),
+            _ => {}
         }
 
-        debug!("type_segment={:?} fn_segment={:?}", type_segment, fn_segment);
-
         // Now that we have categorized what space the parameters for each
         // segment belong to, let's sort out the parameters that the user
         // provided (if any) into their appropriate spaces. We'll also report
         // errors if type parameters are provided in an inappropriate place.
-        let poly_segments = type_segment.is_some() as usize +
-                            fn_segment.is_some() as usize;
-        AstConv::prohibit_generics(self, &segments[..segments.len() - poly_segments]);
+        let mut generic_segs = HashSet::new();
+        for PathSeg(_, index) in &path_segs {
+            generic_segs.insert(index);
+        }
+        let segs: Vec<_> = segments
+            .iter()
+            .enumerate()
+            .filter_map(|(index, seg)| {
+                if !generic_segs.contains(&index) {
+                    Some(seg)
+                } else {
+                    None
+                }
+            })
+            .cloned()
+            .collect();
+        AstConv::prohibit_generics(self, &segs);
```

Comment:
Is this the only use of the variable `segs`? You might want to change `prohibit_generics` to take an iterator or something. Either way, having a variable here that's not `non_generic_segs` is confusing.